### PR TITLE
Fixes #4 - Add support for encrypting using createCipheriv and decryp…

### LINKED ISSUE #4: Add support for encrypting using createCipheriv and decrypting using createDecipheriv
In Node 10, `createCipheriv` and `createDecipheriv` were deprecated. We should change the `encrypt` function to use `createCipheriv` and support a backwards compatible decryption path.

### PR DESCRIPTION
**Pull request by austinkelleher**
…ting using createDecipheriv

### CODE REVIEW

File: src/index.test.ts [hunk 1]
```diff
@@ -59,17 +75,30 @@ test("should be able to encrypt data", async () => {
...
-  } as Encrypted);
+    keyCiphertext: keyCiphertext as unknown as Buffer,
+    dataCiphertext: expect.any(Buffer),
+  };
+
+  expect(encrypted).toEqual(expectedEncryptedResult);
+  initializeMockKmsDecrypt(keyCiphertext);
+
+  const decrypted = await decryptObject({
+    algorithm: "aes256",
```
Comment (philidem) [inline src/index.test.ts]: I had suggested using a new `algorithm` when using an initialization vector. For example, `algorithm: "aes256:iv"` (the `:iv` suffix could be used to indicate that an initialization vector was used).

File: src/index.ts [hunk 2]
```diff
@@ -1,10 +1,52 @@
...
+ *
+ * @param params {DecryptUsingTargetedDecipherMethodParams}
+ */
+async function decryptUsingTargetedDecipherMethod (
+  params: DecryptUsingTargetedDecipherMethodParams
+) {
+  const { algorithm, password, dataCiphertext } = params;
+  const charIndexAfterInitializationVector = INITIALIZATION_VECTOR_SIZE + 1;
+
+  if (dataCiphertext.slice(INITIALIZATION_VECTOR_SIZE, charIndexAfterInitializationVector).toString() === ':') {
```
Comment (philidem) [inline src/index.ts]: Couldn't a `:` end up in the `dataCiphertext` just be chance since it is binary data?

File: src/crypto.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+export function pbkdf2Async(
+  password: crypto.BinaryLike,
+  initializationVector: Buffer
+) {
+  return new Promise<Buffer>((resolve, reject) => {
+    crypto.pbkdf2(
+      password,
+      initializationVector.toString(),
+      10000,
```
Comment (philidem) [inline src/crypto.ts]: What is `10000`? Comment would help here

File: src/crypto.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+      32,
```
Comment (philidem) [inline src/crypto.ts]: What is `32`? Code comment would help here

File: src/index.ts [hunk 5]
```diff
@@ -26,12 +68,16 @@ export async function encrypt(
...
+  const initializationVector = crypto.randomBytes(INITIALIZATION_VECTOR_SIZE);
+
+  const key = await pbkdf2Async(password, initializationVector);
+  const cipher = crypto.createCipheriv(algorithm, key, initializationVector);
 
   const dataCiphertext = Buffer.concat([
-    cipher.update(encryptRequest.data),
-    cipher.final(),
+    initializationVector,
+    Buffer.from(':'),
```
Comment (philidem) [inline src/index.ts]: If you're going to use a separator, pull it out as a constant `const SEPARATOR = Buffer.from(':', 'utf8')`

File: src/index.ts (hunk 2 above)
Comment (philidem) [inline src/index.ts]: You should use `Buffer.from(':', 'utf8').length` to determine length of the separator (that way you don't make any assumptions on the encoding)

**Comment by philidem** [comment]
Overall, I would have preferred introducing a new algorithm name constant (e.g. `aes256:iv`) instead of inspecting the `dataCiphertext` to see if it happens to contain the `":"` separator string at the position beyond the possible initialization vector.

**Comment by austinkelleher** [comment]
Thanks for the great feedback @philidem. I've changed the `encrypt` function to always encrypt using an initialization vector and the `decrypt` function now takes an optional initialization vector. I think this simplifies the process quite a bit.

**Pull request merged**